sys/kern/tty.c

```diff
@@ -493,17 +495,28 @@ void tty_input(tty_t *tty, uint8_t c) {
         tty_output(tty, '\b');
     }
     tty_notify_out(tty);
-    return;
+    return true;
   } else {
     /* Raw (non-canonical) mode */
     if (tty->t_inq.count >= TTY_QUEUE_SIZE) {
-      tty_bell(tty);
-      return;
+      tty_in_hiwat(tty);
+      return false;
     }
 
     ringbuf_putb(&tty->t_inq, c);
     tty_wakeup(tty);
+    return true;
+  }
+}
+
+static void tty_check_in_lowat(tty_t *tty) {
+  if (!(tty->t_flags & TF_IN_HIWAT))
     return;
+
+  if (tty->t_inq.count < TTY_IN_LOW_WATER) {
+    tty->t_flags &= ~TF_IN_HIWAT;
+    if (tty->t_ops.t_notify_in)
```

Comment:
Consider introducing `tty` method wrapper as mentioned in the other review.

Reply:
Done